### DISCUSSION TITLE
Issue with [@types/xrm] control with missing methods e.g setNotfication

### DISCUSSION THREAD

**Discussion by westerdaled** [DefinitelyTyped/DefinitelyTyped#67806] [Issues with a @types package]
In my setup from my project 

` npm install @types/xrm --save-dev`

in my package.json
  ```
"devDependencies": {
    "@types/node": "^20.10.3",
    "@types/xrm": "^9.0.77",
    "typescript": "^5.3.2"
```

 Edtiing the TS file in  VSCode 

I want to select` setNotification` in the control
![event handler issue](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/937208/f8064fb9-3a88-48a3-8158-40b9c8c5baee)

Thanks to the community there is a work around 

My original code

  ```
//check DoB
   var dobFieldLogicalName = "test_dateofbirth";
   var dobField = formContext.getAttribute(dobFieldLogicalName);
   var dobFieldControl = formContext.getControl(dobFieldLogicalName);
```
### workaround which gives you the missing methods

  ```
 var dobFieldControl1 = formContext.getControl<Xrm.Controls.DateControl>(dobFieldLogicalName);
   dobFieldControl1.setNotification( "eee","whatever");
```

[a comment by an automated account is not shown]

**Reply by westerdaled** [reply]
Thanks for the heads up bot!

**Comment by daryllabar** [comment]
`setNotification` is not valid on all controls, therefore, you have to cast it to a control type that does implement it in order to use it.  Seems to be 100% correct the way it works now.

**Reply by westerdaled** [reply]
Ok, thanks for the update.  Does the documentation cover the revealed list of methods per control type, when casting?

**Reply by filcole** [reply]
You may have more luck with 

```typescript
const dobFieldControl: Xrm.Controls.DateControl = formContext.getControl(dobFieldLogicalName);
```

or even

```typescript
const dobAttrib: Xrm.Attributes.DateAttribute = formContext.getAttribute(dobFieldLogicalName);
dobAttrib.controls.forEach((c) => c.setDisabled(true));     // Replace setDisabled as necessary
```

The type definitions list the available methods per control - or discover them via IntelliSense within VSCode etc

**Reply by westerdaled** [reply]
It  looks like there are a couple of  ways to make a viable declaration:

```
  var dobFieldControl1 = formContext.getControl<Xrm.Controls.DateControl>(dobFieldLogicalName);
   dobFieldControl1.setNotification( "First syntax","whatever");

   const dobFieldControl2 : Xrm.Controls.DateControl  = formContext.getControl(dobFieldLogicalName);
   dobFieldControl2.setNotification( "Second syntax","whatever");
```
Not sure which format I prefer🤔.

Also the intellisense seems to be limited given the variety of control types I will be using on the form IMHO.

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/937208/5985c47c-965b-49b4-bb9a-4706ba66128e)

**Comment by westerdaled** [comment]
Thanks, I will try the syntax in the morning. I am having JavaScript detox atm🤗

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
